pkg/scheduler/plugins/loadaware/load_aware.go

```diff
@@ -0,0 +1,304 @@
+				return framework.NewStatus(framework.Unschedulable, fmt.Sprintf(ErrReasonUsageExceedThreshold, resourceName))
+			}
+		}
+	}
+
+	return nil
+}
+
+func (p *Plugin) ScoreExtensions() framework.ScoreExtensions {
+	return nil
+}
+
+func (p *Plugin) Reserve(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status {
+	p.podAssignCache.assign(nodeName, pod)
+	return nil
+}
+
+func (p *Plugin) Unreserve(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) {
+	p.podAssignCache.unAssign(nodeName, pod)
+}
+
+func (p *Plugin) Score(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) (int64, *framework.Status) {
+	nodeInfo, err := p.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+	if err != nil {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("getting node %q from Snapshot: %v", nodeName, err))
+	}
+	node := nodeInfo.Node()
+	if node == nil {
+		return 0, framework.NewStatus(framework.Error, "node not found")
+	}
+	nodeMetric, err := p.nodeMetricLister.Get(nodeName)
+	if err != nil {
+		return 0, framework.NewStatus(framework.Error, "nodeMetric not found")
+	}
+	if p.args.NodeMetricExpirationSeconds != nil && isNodeMetricExpired(nodeMetric, *p.args.NodeMetricExpirationSeconds) {
+		return 0, nil
+	}
+
+	estimatedUsed := estimatedPodUsed(pod, p.args.ResourceWeights, p.args.EstimatedScalingFactors)
+	estimatedAssignedPodUsage := p.estimatedAssignedPodUsage(nodeName, nodeMetric)
+	for resourceName, value := range estimatedAssignedPodUsage {
+		estimatedUsed[resourceName] += value
+	}
+
+	allocatable := make(map[corev1.ResourceName]int64)
+	for resourceName := range p.args.ResourceWeights {
+		quantity := node.Status.Allocatable[resourceName]
+		if resourceName == corev1.ResourceCPU {
+			allocatable[resourceName] = quantity.MilliValue()
+		} else {
+			allocatable[resourceName] = quantity.Value()
+		}
+		if nodeMetric.Status.NodeMetric != nil {
+			quantity = nodeMetric.Status.NodeMetric.NodeUsage.ResourceList[resourceName]
+			if resourceName == corev1.ResourceCPU {
+				estimatedUsed[resourceName] += quantity.MilliValue()
+			} else {
+				estimatedUsed[resourceName] += quantity.Value()
+			}
+		}
+	}
+
+	score := loadAwareSchedulingScorer(p.args.ResourceWeights, estimatedUsed, allocatable)
+	return score, nil
+}
+
+func isNodeMetricExpired(nodeMetric *slov1alpha1.NodeMetric, nodeMetricExpirationSeconds int64) bool {
+	return nodeMetric == nil ||
+		nodeMetric.Status.UpdateTime == nil ||
+		nodeMetricExpirationSeconds > 0 &&
+			time.Since(nodeMetric.Status.UpdateTime.Time) >= time.Duration(nodeMetricExpirationSeconds)*time.Second
+}
+
+func (p *Plugin) estimatedAssignedPodUsage(nodeName string, nodeMetric *slov1alpha1.NodeMetric) map[corev1.ResourceName]int64 {
+	estimatedUsed := make(map[corev1.ResourceName]int64)
+	nodeMetricReportInterval := getNodeMetricReportInterval(nodeMetric)
+	p.podAssignCache.lock.RLock()
```

Comment:
Is it possible to reduce the lock granularity as there is no correlation among different nodes?

Reply:
ok